### DISCUSSION TITLE
HTML Img tag not parsed if contained in a text string

### DISCUSSION THREAD

**Discussion by Chetral** [evennia/evennia#3538] [Evennia Engine Development]
I'm trying to add, for every objects and mobs, an icon, passed as a `<img src="xxx">` tag.
If i send the tag alone with `caller.msg(html=tag_variable)` it returns the image, if the variable contains also other texts, all the string is returnet as normal text and in the output i see:
`<img src="xxx">` A torch is here.

Doing some tests in `py`, if i pass the string in parse_html from text2html module and then send it in `caller.msg(html=tag_variable)` , it returns the correct image BUT stops to parse the ANSI color tag, if i do the same and send it as `caller.msg(text=tag_variable)` i got the color parsed, but not the img.

**Comment by InspectorCaracal** [comment]
I haven't much used the html kwarg, but I believe what you want is to have those two things be separate.

Example:
```python
caller.msg(html=img_element)
caller.msg(text=text_message)
```

It may also be possible to do `caller.msg(text=text_message, html=img_element)` with one call to msg, but the important part is that your image needs to be given to the html keyword and your regular text needs to be given to the text keyword, they cannot both be in the same keyword since they function differently.

**Reply by Chetral** [reply]
Hi inspector (i'm Wabbit in the discord, btw... didn't noticed here i have different nick :D)
The text2html module extract a link `<a href="something">Click Here</a>` even if it is in a string.

I'm using the normal `return_apparence` func from room object, when creating the list of objects in the room with `get_display_things` i add the html tag in the name of the object (via `get_numbered_name` func).
In all this process the only change i made is in `get_numbered_name` func. At return_apparance arrives a single long string, and that is passed to the CmdLook and then to caller.msg, where i tried the `caller.msg(html=room.return_apparence(caller))`.
To split the message as your example i will have to rewrite a big part of the CmdLook and, more important, i think caller.msg create a new line every time it is called, the icon will result on a different line.

**Reply by InspectorCaracal** [reply]
Hmmm.... I don't think your idea of just sticking the img element directly into the pre-parsing string in `get_numbered_name` is going to work very well, due to the order of the various parsing and formatting steps.

I don't know if this is the *best* way to do it, but if I were trying to achieve this effect, I would actually create a new Evennia-style markup for images. Like how we have `|lclook object|ltobject|le` to generate clickable text, you could add e.g. an `|im` tag, which would work somthing like `|imurl/for/image.png|im` and then be added into the parsers. That's quite a lot of work, though, and I probably not what you signed up for. 😆

It *does* make me want to support it in core, though. Hmmmmmmmmmm.

**Reply by Chetral** [reply]
This would meant override the text2html module to add a parser for the image...

But since you are going to put it in the core, I don't need to do it 😬

For what I understood, one of the sub in the parser search for html tags, but the regex (that for me is something related to magic) search also for a endline. But the endline are replaced in the first step with the `<br>` for web page. 
Another sub instead searches directly for `<a ...> some text </a>` and this works. Now that I'm thinking at it could be an easy thing: or the order are effectively wrong and the general html search is not working, or I simply need to add a new regex for img tags. But now I'm writing from the phone so I will check tomorrow after work

**Comment by Chetral** [comment]
I've modified the text2html module (copied out of evennia root and modified), I created a tag for the image and added it to the regex.
This is what i see if i use `caller.msg(html= parse_html(room.return_appearance(caller))` <--- the parse_html is my version
<img width="545" alt="image" src="https://github.com/evennia/evennia/assets/55252937/ba74270b-6dcb-4036-af18-3ad15f897329">
Seems to not working... BUT!!
If i open `py`, import my parse and send the message:
<img width="920" alt="image" src="https://github.com/evennia/evennia/assets/55252937/4484c2d0-e351-43ab-b21a-e761ed3df3f3">
The icon is there!
What difference there is between the server sending the message and me doing it manually?

**Reply by InspectorCaracal** [reply]
Since it seems like the part it's failing at on the server is parsing the inputs into HTML, I recommend putting in some debug logging to see what values are getting passed to where to find out what the issue is.

**Comment by Chetral** [comment]
GOT IT! And i want to cry.
First error: in the character CmdSet there is an override version of the CmdLook... but seems not "all" override. If i type `l` or `look` it gets the standard version, so the description is not sent as `html =` but as `text =`. If i type `g` or `guarda` (that are aliases only present in the new CmdLook) the icon is parsed correctly.
Second error: the `at_post_move` hook does not anymore force an `execute_cmd`, but return a msg (txt) of the description.
